fix: Address disruptive UI bugs

dcalhoun opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Address styling bugs that disrupt the UX. This reintroduces some hover and focus styles included from the web editor, this was done as the one-off overrides resulted in inconsistent styling and regressions elsewhere[^1]. We should consider modifying and improving button styles in a more robust manner at a later time. 

It's worth noting that this does not fix all style issues, only those I deemed particularly disruptive. 

* Increased editor toolbar button tap target size. 
* Fixed illegible buttons caused from lingering hover styles. 
* Fixed misaligned inserter tab. 
* Fixed inconsistencies in theme styles application—e.g., text size, text color, horizontal root padding. 
* Fixed erroneous default browser styles on `fieldset` elements in block settings—e.g., Image block border radius. 
* Refactored stylesheets to align with Gutenberg project practices—e.g., leveraging Sass, co-locating related styles and markup. 

[^1]: E.g., tapping the buttons in an Image block's placeholder could leave the button styles in odd states, with missing borders or illegible text. 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Fixes https://github.com/Automattic/dotcom-forge/issues/8678. Fixes https://github.com/Automattic/dotcom-forge/issues/9943. 

The bugs degrade the UX and erode user trust in the product.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Refactor to further utilize Gutenberg project components and styling practices,
benefiting from well-tested UI and modularizing custom styles.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
Navigate the editor, noting any unexpected styles, layout issues, or bugs.

### Accessibility Testing Instructions
<!-- How can you test the changes by using a keyboard/screen reader only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
Navigate the editor, noting any unexpected styles, layout issues, or bugs.

## Screenshots or screencast <!-- if applicable -->
There are numerous visual fixes, too many to capture in screenshots.

dcalhoun committed:
build: Fix package version and metadata details

dcalhoun committed:
style: Co-locate import statements

Follow Gutenberg project practices.

dcalhoun committed:
refactor: Add Default site API root negating conditionals

An empty string avoids errors previously guarded by conditionals.

dcalhoun committed:
refactor: Rely on default styles provided by API

Avoid duplicative twice-loaded default editor styles.

dcalhoun committed:
build: Configure Sass compilation

Align with Gutenberg project styling practices.

dcalhoun committed:
build: Upgrade @vitejs/plugin-react

Necessary to address peer dependency conflicts, likely from upgrading
Vite.

dcalhoun committed:
build: Add clsx

dcalhoun committed:
refactor: Co-locate and reduce editor and toolbar styles

Utilize Gutenberg UI components to fix style bugs.

dcalhoun committed:
refactor: Co-locate inserter button styles

dcalhoun committed:
refactor: Remove unused toolbar styles

dcalhoun committed:
refactor: Remove unused react-modal-sheet styles

dcalhoun committed:
refactor: Co-locate editor load notice styles

dcalhoun committed:
refactor: Co-locate editor interface padding styles

dcalhoun committed:
fix: Remove default block toolbar box shadow

The shadow is unnecessary as we do not render the block toolbar as a
popover like the web.

dcalhoun committed:
refactor: Co-locate disabled default tap feedback styles

dcalhoun committed:
refactor: Co-locate button focus styles

dcalhoun committed:
style: Alphabetize CSS attributes

dcalhoun committed:
refactor: Remove default "theme" styles

These appear unnecessary now that editor styles are applied elsewhere.

dcalhoun committed:
refactor: Remove unused rich text and popover overrides

These appear to have no impact.

dcalhoun committed:
fix: Disable block toolbar drag handle

This handle is irrelevant for a fixed-position toolbar.

dcalhoun committed:
refactor: Co-locate editor snackbar styles

dcalhoun committed:
fix: Reinstate full-height block canvas

dcalhoun committed:
fix: Disable button hover styles

The hover style is irrelevant for touch devices, and its presence
lingers when tapping interactive elements in mobile browsers.

dcalhoun committed:
fix: Disable block toolbar scrolling

Rely upon editor toolbar scrolling rather than block toolbar scrolling
to avoid nested scrolling elements.

dcalhoun committed:
feat: Hide the editor toolbar scrollbar

Hide the block toolbar scrolling to avoid accidental interactions with
the scrollbar when swiping the short editor toolbar.

dcalhoun committed:
feat: Style UI root

Align root styles with Gutenberg editor.

dcalhoun committed:
feat: Refine snackbar positioning

Balance space beneath and beside snackbars.

dcalhoun committed:
fix: Remove problematic style overrides

These overrides introduce subtle bugs with button states. E.g., some
buttons utilize an `aria-pressed` state applying a background color,
which often collides with the inherited color; some stateless button
variant styles rely include outlines, which are removed by the stateful
style reset.

dcalhoun committed:
fix: Remove problematic image width binding

The low specificity of this selector resulted in broken styles for
buttons within media block placeholder UI.

dcalhoun committed:
fix: Avoid global button overrides

These prove problematic and the result is inconsistent across button
states.

dcalhoun committed:
refactor: Editor dimensions strategy mirrors web editor

Reduce the likelihood of nuanced styling issues by aligning with the web
editor.

dcalhoun committed:
fix: Reset fieldset browser styles

Gutenberg relies upon styles provided by the WP Admin context, which are
absent from the GutenbergKit editor environment. Loading the entirety of
the `common.css` stylesheet creates style collisions, so we manually
load this one `fieldset` styling.

dcalhoun committed:
refactor: Remove border radius overrides

Embrace Gutenberg's default styling, at least until we override the
styles in a more wholistic and maintainable manner.

dcalhoun committed:
refactor: Remove now unused toolbar styles

This became unused in a previous refactor to utilize core Gutenberg
components.

dcalhoun committed:
refactor: Remove redundant toolbar group border

Replaced by usage of `ToolbarGroup` component within the editor toolbar.

dcalhoun committed:
fix: Prevent top-level editor focus outline

Without this, selecting multiple blocks results in a unexpected focus
outline around the top-level editor container. This outline does not
occur in the web editor, as the iframed editor renders as `body`
element--which user agent styles (browser defaults) removes the outline
from this element.

dcalhoun committed:
fix: Conditionally apply default theme styles

These styles overrode theme styles, resulting in an unexpected and
inconsistent application of theme styles.

dcalhoun committed:
refactor: Disable erroneous unresolved import error

The linter's module resolution appears to be disrupted by Vite's query
parameter for inline CSS values.

dcalhoun committed:
fix: Apply default styles when the theme styles feature flag is disabled

Previously, the default theme styles were only applied when the editor
settings were missing. This is no the only scenario when they should be
applied. They should also be applied if the user has disabled the
experimental "theme styles" feature flag.

dcalhoun committed:
fix: Ensure the editor toolbar is above the block canvas

dcalhoun committed:
refactor: Co-locate block toolbar styles with the editor toolbar

The block toolbar is rendered within the editor toolbar.

dcalhoun committed:
fix: Force a full-width inserter tabs

The Patterns/Media tabs rendered with a width smaller than the overall
popover container, leaving an odd layout.

dcalhoun committed:
fix: Ensure theme padding is applied to avoid cut-off content

Without using the theme's root padding, content can be cut off when
using larger text sizes.

dcalhoun committed:
fix: Ensure root paddings respect the "theme styles" feature flag

dcalhoun committed:
fix: Reinstate hiding of block popover

Prevent the popover from stealing focus away from the inserter, which
results in needing to tap a block type twice to insert it.

dcalhoun committed:
fix: Ensure toolbar sits atop image resize handles within canvas

dcalhoun commented on src/index.scss:
The lines removed were either relocated to a specific component folder or removed due to problematic hover style overrides.

kean commented:
> Increased editor toolbar button tap target size.

The buttons initially appeared too large, but after testing on a physical device, I thought they were just the right size.

> Fixed illegible buttons caused from lingering hover styles.

Confirmed as fixed.

> Fixed misaligned inserter tab.

Also fixed 👏 

There is some misalignment with popovers (no right margin):

<img src="https://github.com/user-attachments/assets/ea190182-963c-47ba-8502-de2bf229b935" width="320px">

> Refactored stylesheets to align with Gutenberg project practices

Some styles appear to be lost. For example, menus now have small buttons again. I completely recognize that initial implementation was not sustainable to say the least 😅 

In terms of the styling in general, I don't think we need to over-complicate it, but it needs some changes as the current styles are not designed for mobile. It can be as simple as:

- Use Apple's color palette. For example, separators and borders should be light gray. The standard Gutenberg black borders are too harsh.
- Add ~8pt corner radius to menus, and anywhere they make sense.
- Use standard body fonts (17pt San Francisco) for all menus, labels, etc.

This will get us 90% to where it feels nice and native.

 For reference, the original design looked more like this:

<img width="320" alt="screenshot-2024-07-01-at-11 13 36e280afam-1" src="https://github.com/user-attachments/assets/5361dea1-122b-4587-bc45-5101833a2b27" />

kean reviewed (approved):
LGTM, minus the missing styles (see the previous comment)

dcalhoun commented:
Thank you for your testing, review, and thoughts, @kean. I really appreciate it. 🙇🏻‍♂️ 

> Some styles appear to be lost. For example, menus now have small buttons again. 

Will you share a specific example of a "small button" please? That will help ensure I do not misunderstand your discovery. 

> Use Apple's color palette. For example, separators and borders should be light gray. The standard Gutenberg black borders are too harsh.

It appears I introduced darker borders by wrapping our editor toolbar with `Toolbar`. The dropdown menus in the web editor header toolbar do _not_ have the dark borders, but utilize the lighter gray. I'll investigate a solution. 

That said, at this time, I am not in complete agreement that applying Apple's color palette is the most appropriate solution—at least not everywhere, and not without a robust solution that avoids regressions from CSS overrides. 

1. **We should consider the Android platform.** We could accomplish with platform-specific, top-level classes on the root element for diverging styles, but it's worth debating if that is worth the maintenance burden. 
1. **There is the "uncanny valley" problem.** Attempts to shed Gutenberg's unique web styles and align with Apple's could result in degraded UX where we mimic native UI, but do not deliver the same quality with our web UI. In some ways, a perception or expectation issue. 
1. **Simple CSS overrides are brittle.** Part of the focus of these proposed changes is reverting overrides to ensure the UI is at least consistent within itself, albeit inconsistent with the host platform. Long term, if we truly want to pursue unique platform-aligned styles, we should likely build custom web components for the mobile platform—i.e., avoid the pattern of reusing web UI components and altering them with CSS. 

> Add ~8pt corner radius to menus, and anywhere they make sense.

I'm not inherently opposed to this, but it relates to what I shared above regarding border colors and platform alignment. There are instances where border radii were modified, but introduced regressions—e.g., mismatched border radii between parent and child. 

> Use standard body fonts (17pt San Francisco) for all menus, labels, etc.

Will you share a specific example that you believe causes problems currently please? I do not want to misinterpret your statement.

kean commented:
> Attempts to shed Gutenberg's unique web styles and align with Apple's could result in degraded UX where we mimic native.

Sorry, I wasn't entirely clear as that was not my suggestion. My point is that Gutenberg isn't optimized for mobile devices, and it's our job to do it. I think it should retain its unique WordPress look and feel. It's really the only option.

The primary point is that text should be legible. Mobile requires much larger fonts than desktop. On macOS, body (most text) uses size 13pt. The equivalent on iOS is 17pt. It's pretty clear that the current fonts are too small if you contrast it with the native components:

<img width="1618" alt="Screenshot 2024-12-20 at 9 24 42 AM" src="https://github.com/user-attachments/assets/3352ea23-cd57-4770-96fd-98177c261325" />

For Gutenberg menus, I would probably go a little bit smaller with 16pt as it typically provides a lot of options. But, ideally, these need to be grouped and hidden in submenus on mobile where applicable.

I also advocate for using light gray colors where applicable for borders and adding a corner radius on iOS. It will make it feel more like part of the app. Rounded corners are quintessentially Apple, and it will go a long way to make it feel more like also part of the platform. I hope it won't be complicated from the technical point.

dcalhoun committed:
fix: Use the "unstyled" toolbar variant

Avoid application of the darker border styles intended for the block
toolbar that hovers above a block. The default "unstyled" variant is
intended for other uses, like a fixed-position editor toolbar.

dcalhoun committed:
feat: Adjust editor toolbar tap target size

Slightly reduce to the minimum recommended 44pt, reduce toolbar
min-height to match.

dcalhoun committed:
feat: Increase baseline text size for buttons

Improve legibility and tap target sizes.

dcalhoun committed:
feat: Increase menu item touch target size

Mitigate hit slop with a taller touch target. Use a slightly smaller
height that the recommended to balance touch target size with
overflowing menu lists. Also, menu items are generally wider, making
taping them easier.

dcalhoun committed:
feat: Increase touch target size for auto-complete items

dcalhoun commented:
Before merging, I pushed a few changes improving legibility and tap targets. We can iterate further in future pull requests. 

* Fixed the dark gray borders found in the editor toolbar dropdown menus. 
* Reduced the editor toolbar button tap targets from 48pt to 46pt. 
* Increased the baseline text for interactive buttons from 13pt to 17pt. 
* Increased the minimum tap target height for menu list and auto-complete items from 36pt to 42pt.

kean commented:
Awesome! Let's ship it in 25.6 :shipit:
